### PR TITLE
Ui

### PR DESCRIPTION
**Pull request by lazukr**
please check if the usage of your hooks are as intended and advise any changes that you think might work better.

### CODE REVIEW

File: game.js [hunk 1]
```diff
@@ -280,7 +280,7 @@ class Game {
     this.tickSize = 40;
     this.generation = 0;
     this.eventListeners = new EventListeners();
-
+    this.isRunning = false;
```
Comment (jsatgit) [inline game.js]: why did you add this here? The reason why I added this to the reset function is so that you don't have to initialize the value twice.

File: game.js [hunk 2]
```diff
@@ -335,9 +335,11 @@ class Game {
   reset() {
     this.view = new View(this.squareSize);
     this.board = new Board(this.width, this.height);
-    this.isRunning = true;
-
+    this.isRunning = false;
+    this.generation = 0;
```
Comment (jsatgit) [inline game.js:71]: good call on reseting the generation here. but make sure to also remove it from the game constructor. the purpose of calling the reset function in the constructor is so that you don't have to set it in two places.

File: game.js [hunk 3]
```diff
@@ -335,9 +335,11 @@ class Game {
...
     this._addArmiesToBoard();
+    this.render();
```
Comment (jsatgit) [inline game.js]: don't call the `render` in `reset`. `reset` means reset the game state and nothing else

File: game.js [hunk 4]
```diff
@@ -335,9 +335,11 @@ class Game {
...
+    this.eventListeners.dispatch(Events.NEW_GENERATION, this.generation);
```
Comment (jsatgit) [inline game.js]: don't dispatch events here. also the generation is 0. The purpose of events is to notify when there is a change. There has been no change, you know the initial state is generation = 0

File: game.js [hunk 5]
```diff
@@ -380,11 +387,22 @@ function main() {
...
-    height: 25,
-    squareSize: 20,
+    width: WIDTH,
+    height: HEIGHT,
+    squareSize: SQUARE_SIZE,
     players: [player1, player2],
   });
 
-  game.start();
+  const table = document.getElementById("table-stats");
```
Comment (jsatgit) [inline game.js]: i shouldn't be seeing any `document.getElementById` calls here. That is way too low level. Please work on the same abstraction layer

File: game.js [hunk 6]
```diff
@@ -369,6 +371,11 @@ class Game {
 }
 
 function main() {
+
+  const WIDTH = 25;
```
Comment (jsatgit) [inline game.js]: constants should live on the module level, not a local variable - otherwise they would just be a local variable

File: game.js [hunk 7]
```diff
@@ -380,11 +387,22 @@ function main() {
...
+    squareSize: SQUARE_SIZE,
     players: [player1, player2],
   });
 
-  game.start();
+  const table = document.getElementById("table-stats");
+  table.width = WIDTH*SQUARE_SIZE;
+  table.height = HEIGHT*SQUARE_SIZE;
+
+  const ui = new UI({
```
Comment (jsatgit) [inline game.js]: why does the UI take so many parameters? As a user of the UI i should just be able to construct it as easily as possible. Why do I need to know the playButton's element id?

File: game.js [hunk 8]
```diff
@@ -380,11 +387,22 @@ function main() {
...
+    game: game,
+    playButton: document.getElementById("play-button"),
+    stopButton: document.getElementById("stop-button"),
+    stepButton: document.getElementById("step-button"),
+    generationCounter: document.getElementById("generation-counter"),
+    events: Events,
```
Comment (jsatgit) [inline game.js]: why are you passing a module level enum as a configuration parameter to ur ui? u should just import that

File: game.js [hunk 9]
```diff
@@ -380,11 +387,22 @@ function main() {
...
-    squareSize: 20,
+    width: WIDTH,
+    height: HEIGHT,
+    squareSize: SQUARE_SIZE,
     players: [player1, player2],
   });
 
-  game.start();
+  const table = document.getElementById("table-stats");
+  table.width = WIDTH*SQUARE_SIZE;
```
Comment (jsatgit) [inline game.js]: I'm not sure why we are constructing a table here and giving it properties of the game. If table is a specific component of the ui, then it should be constructed within the ui. If the table needs to be constructed based on game configs then it should ask the game for those configs

File: styles.css [hunk 10]
```diff
@@ -0,0 +1,22 @@
+body {
+
```
Comment (jsatgit) [inline styles.css]: there is some really weird spacing here

File: ui.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    this.stopButton = config.stopButton;
+    this.generationCounter = config.generationCounter;
+    this.events = config.events;
+    //this.mousePosition = config.mousePosition;
+    this.initUI();
+  }
+
+  initUI() {
+    this.playIcon = this.playButton.firstElementChild;
+    this.stopButton.onclick = () => {
```
Comment (jsatgit) [inline ui.js]: please don't register handlers like this. use the element's addEventListener

File: ui.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
+
+const PLAY_ICON = "play_arrow";
+const PAUSE_ICON = "pause";
+
+class UI {
+  constructor(config) {
+    this.game = config.game;
+    this.playButton = config.playButton;
```
Comment (jsatgit) [inline ui.js]: there is no need for these to be configurable

File: ui.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    this.stepButton = config.stepButton;
+    this.stopButton = config.stopButton;
+    this.generationCounter = config.generationCounter;
+    this.events = config.events;
+    //this.mousePosition = config.mousePosition;
```
Comment (jsatgit) [inline ui.js]: comments like these should be deleted

File: ui.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    this.initUI();
+  }
+
+  initUI() {
```
Comment (jsatgit) [inline ui.js]: you could be more specific and call these set of actions as registering event handlers

File: ui.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    this.playIcon = this.playButton.firstElementChild;
+    this.stopButton.onclick = () => {
+      this.game.reset();
+    };
+    this.playButton.onclick = () => {
```
Comment (jsatgit) [inline ui.js]: play button should not be a play button some of the time and a pause button the other part of the time. you called it a play button so it should just be a play button

File: ui.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+      if (this.game.isRunning) {
+        this.game.pause();
+        this.playIcon.innerHTML = PLAY_ICON;
+      } else {
+        this.game.start();
+        this.playIcon.innerHTML = PAUSE_ICON;
+      }
+    };
+    this.stepButton.onclick = () => {
+      this.playIcon.innerHTML = PLAY_ICON;
```
Comment (jsatgit) [inline ui.js]: why does the step button need any knowledge of the play icon?

**Review by jsatgit — CHANGES_REQUESTED** [review]

File: game.js (hunk 4 above)
Comment (lazukr) [reply game.js]: This was done because when you reset, it doesn't dispatch an event to update the generation counter back to 0 (Also the reason for `this.render`. Namely, if the generation is at a non zero value, and I click reset, it will still display that non-zero value despite the fact that the game has been reset. I'll look for a different way to ensure it is consistent when reset occurs.

**Review by jsatgit — APPROVED** [review]
